polkadot/runtime/parachains/src/coretime/mod.rs

```diff
@@ -244,11 +262,55 @@ impl<T: Config> Pallet<T> {
 				Location::new(0, [Junction::Parachain(T::BrokerId::get())]),
 				message,
 			) {
-				log::error!("Sending `NotifyCoreCount` to coretime chain failed: {:?}", err);
+				log::error!(target: LOG_TARGET, "Sending `NotifyCoreCount` to coretime chain failed: {:?}", err);
 			}
 		}
 	}
 
+	/// Provide the amount of revenue accumulated from Instantaneous Coretime Sales from Relay-chain
+	/// block number last_until to until, not including until itself. last_until is defined as being
+	/// the until argument of the last notify_revenue message sent, or zero for the first call. If
+	/// revenue is None, this indicates that the information is no longer available. This explicitly
+	/// disregards the possibility of multiple parachains requesting and being notified of revenue
+	/// information.
+	///
+	/// The Relay-chain must be configured to ensure that only a single revenue information
+	/// destination exists.
+	pub fn notify_revenue(when: BlockNumber) -> DispatchResult {
+		let now = <frame_system::Pallet<T>>::block_number();
+		let when_bnf: BlockNumberFor<T> = when.into();
+
+		// When cannot be in the future.
+		ensure!(when_bnf <= now, Error::<T>::RequestedFutureRevenue);
+
+		let revenue = <assigner_on_demand::Pallet<T>>::revenue_until(when_bnf);
+		log::info!(target: LOG_TARGET, "Revenue info requested: {:?}", revenue);
+		match TryInto::<Balance>::try_into(revenue) {
+			Ok(raw_revenue) => {
+				log::info!(target: LOG_TARGET, "Revenue into balance success: {:?}", raw_revenue);
+				let message = Xcm(vec![
+					Instruction::UnpaidExecution {
+						weight_limit: WeightLimit::Unlimited,
+						check_origin: None,
+					},
+					mk_coretime_call::<T>(CoretimeCalls::NotifyRevenue((when, raw_revenue))),
```

Comment:
We will also need an accompanying teleport.

Reply:
I definitely need some guidance here. Like, what is a teleport, what should it accompany, and how to implement it :slightly_smiling_face:

Reply:
Well, I took a crash course on assets, currencies, and teleports, and it looks like I'm starting to understand what it all is about. Still, I'm not sure we need that full-scale XCM teleport we were talking about.

The relay chain withdraws the funds for on-demand purchases from the buyer's account and deposits them to the pallet's account. At the same time, it accounts for the revenue (that accounting is just a number not touching economic stuff yet).

Then, the Coretime chain sends a request to the relay chain to provide the revenue info. At that moment, the relay chain burns funds from the pallet account and provides the exact amount of the revenue collected to the Coretime chain.

Thus, when receiving the answer from the relay chain:
1) The Coretime chain already knows the exact amount of revenue;
2) That amount has already been burnt on the relay chain.

It seems that the only thing that should be done here is to mint the amount in question into the Coretime pallet's account. No need for an additional XCM instruction as all the information required is already known to the receiver and is trusted.

Do I miss something?

Reply:
No. Burn -> mint is a teleport. If you think using the actual XCM is overkill here, that is totally fine. There is a difference later once we have credits as then we don't need to teleport anything, but we can deal with that once we are there.